MoreLinq/ToArrayByIndex.cs

```diff
@@ -0,0 +1,241 @@
+#region License and Terms
+// MoreLINQ - Extensions to LINQ to Objects
+// Copyright (c) 2016 Atif Aziz. All rights reserved.
+// 
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+// 
+//     http://www.apache.org/licenses/LICENSE-2.0
+// 
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#endregion
+
+namespace MoreLinq
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Linq;
+
+    static partial class MoreEnumerable
+    {
+        /// <summary>
+        /// Creates an array from an <see cref="IEnumerable{T}"/> where a
+        /// function is used to determine the index at which an element will
+        /// be placed in the array.
+        /// </summary>
+        /// <param name="source">The source sequence for the array.</param>
+        /// <param name="indexSelector">
+        /// A function that maps an element to its index.</param>
+        /// <typeparam name="T">
+        /// The type of the element in<paramref name="source"/>.</typeparam>
+        /// <returns>
+        /// An array that contains the elements from the input sequence. The
+        /// size of the array will be as large as the highest index returned
+        /// by the <paramref name="indexSelector"/> plus 1.
+        /// </returns>
+        /// <remarks>
+        /// This method forces immediate query evaluation. It should not be
+        /// used on infinite sequences. If more than one element maps to the
+        /// same index then the latter element overwrites the former in the
+        /// resulting array.
```

Comment:
Unfortunately when I read it back, it says something else. It says the _last in the whole sequence_ and not _last in the sequence that mapped to the same index_. To avoid ambiguity about how to interpret _last_, it's important to re-iterate that it has to be one mapping to the same index.

I think let's go with the original for now. If someone comes back confused then we can revisit.